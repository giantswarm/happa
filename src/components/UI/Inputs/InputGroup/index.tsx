import { FormField, Text } from 'grommet';
import PropTypes from 'prop-types';
import * as React from 'react';

interface IInputGroupProps
  extends React.ComponentPropsWithoutRef<typeof FormField> {}

const InputGroup = React.forwardRef<HTMLDivElement, IInputGroupProps>(
  ({ label, contentProps, ...props }, ref) => {
    let labelComponent: React.ReactNode = label;
    if (typeof label === 'string') {
      labelComponent = <Text size='large'>{label}</Text>;
    }

    return (
      <FormField
        label={labelComponent}
        margin={{ bottom: 'none' }}
        {...props}
        contentProps={{
          border: false,
          ...contentProps,
        }}
        ref={ref as React.Ref<HTMLDivElement> & string}
      />
    );
  }
);

InputGroup.propTypes = {
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  contentProps: PropTypes.object,
};

export default InputGroup;
