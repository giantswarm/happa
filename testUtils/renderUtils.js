import { render } from '@testing-library/react';
import App from 'App';
import { ConnectedRouter } from 'connected-react-router';
import { ThemeProvider } from 'emotion-theming';
import { createMemoryHistory } from 'history';
import React from 'react';
import { Provider } from 'react-redux';
import { AppRoutes } from 'shared/constants/routes';
import configureStore from 'stores/configureStore';
import theme from 'styles/theme';

export const initialStorage = {
  user:
    '{"email":"developer@giantswarm.io","auth":{"scheme":"giantswarm","token":"a-valid-token"},"isAdmin":true}',
};

/**
 * This function will render the whole app with a mocked store in the route
 * provided.
 * @param {String} route.
 * @param {HTMLElement} container An HTMLElement.
 */
export function renderRouteWithStore(
  initialRoute = AppRoutes.Home,
  state = {},
  storage = initialStorage
) {
  localStorage.replaceWith(storage);

  const history = createMemoryHistory({
    initialEntries: [initialRoute],
    initialIndex: 0,
  });

  const store = configureStore(state, history);

  const app = render(<App {...{ store, theme, history }} />);

  return app;
}

/**
 * Render component in the Theme context
 * @param {React.ReactType} component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {<Q extends TestingLibrary.Queries>TestingLibrary.RenderOptions<Q>} [options] Testing library render options
 */
export function renderWithTheme(component, props, options) {
  return render(getComponentWithTheme(component, props), options);
}

/**
 * Include component in the Theme context and return it
 * @param {React.ReactType} Component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 */
export function getComponentWithTheme(Component, props) {
  return (
    <ThemeProvider theme={theme}>
      <Component {...props} />
    </ThemeProvider>
  );
}

/**
 * Render component in the Store context
 * @param {React.ReactType} component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {Record<string, any>} state Current Store state
 * @param {<Q extends TestingLibrary.Queries>TestingLibrary.RenderOptions<Q>} [options] Testing library render options
 */
export function renderWithStore(component, props, state, options) {
  return render(getComponentWithStore(component, props, state), options);
}

/**
 * Include component in the Store context and return it
 * @param {React.ReactType} Component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {Record<string, any>} state Current Store state
 * @param {Record<string, any>} storage Current LocalStorage value
 * @param {History} history Current Browser history
 */
export function getComponentWithStore(
  Component,
  props = {},
  state = {},
  storage = initialStorage,
  history = createMemoryHistory()
) {
  localStorage.replaceWith(storage);

  const store = configureStore(state, history);

  const app = (
    <Provider store={store}>
      <ThemeProvider theme={theme}>
        <ConnectedRouter history={history}>
          <Component {...props} />
        </ConnectedRouter>
      </ThemeProvider>
    </Provider>
  );

  return app;
}
