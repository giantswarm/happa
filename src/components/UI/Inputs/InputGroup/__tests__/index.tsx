import { renderWithTheme } from 'test/renderUtils';

import InputGroup from '..';

describe('InputGroup', () => {
  it('renders a simple input', () => {
    const { container } = renderWithTheme(InputGroup, {
      label: 'A nice group of inputs',
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
