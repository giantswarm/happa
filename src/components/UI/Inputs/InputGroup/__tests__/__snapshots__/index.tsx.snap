// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputGroup renders a simple input 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 fEWbpp"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 bmLMm"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 iOFkGP FormField__FormFieldBox-sc-m9hood-0 chvcoo"
    >
      <label
        class="StyledText-sc-1sadyjn-0 gGKbiL"
      >
        <span
          class="StyledText-sc-1sadyjn-0 eHHJtv"
        >
          A nice group of inputs
        </span>
      </label>
      <div
        class="StyledBox-sc-13pk1d4-0 bfCPCc FormField__FormFieldContentBox-sc-m9hood-1"
      />
    </div>
  </div>
</div>
`;
